mlir/lib/Dialect/Rock/Transforms/LowerRockReduce.cpp

```diff
@@ -0,0 +1,233 @@
+};
+} // end namespace
+
+// This function will create a view of the tensor that is being
+// reduced from arbitary dimension sizes to bid x tid x iter space.
+static Value createThreadView(Value redInput, int64_t blockSize,
+                              int64_t gridSize, Location loc,
+                              PatternRewriter &rewriter) {
+  auto totalThreads = gridSize * blockSize;
+  auto inpShape = redInput.getType().cast<ShapedType>();
+  auto elementCount = inpShape.getNumElements();
+  auto dataPerThread = (elementCount + (totalThreads - 1)) / totalThreads;
+
+  llvm::errs() << "totalThreads = " << totalThreads << "\n";
+  llvm::errs() << "elementCount = " << elementCount << "\n";
+  llvm::errs() << "dataPerThread = " << dataPerThread << "\n";
+
+  SmallVector<SmallString<8>> topNames;
+  SmallVector<int64_t> topSizes;
+  int64_t topSizeProduct = 1;
+  SmallVector<unsigned int> topDims;
+  SmallVector<StringRef> topNameRefs;
+  for (auto dimAndSize : llvm::enumerate(inpShape.getShape())) {
+    auto dim = dimAndSize.index();
+    auto dimSize = dimAndSize.value();
+    SmallString<8> mname(Twine("m" + Twine(dim)).str());
+    topNames.push_back(mname);
+    topNameRefs.push_back(topNames.back());
+    topSizes.push_back(dimSize);
+    topSizeProduct *= dimSize;
+    topDims.push_back(dim);
+  }
+
+  BottomUpTMBuilder threadsToInpTensor(rewriter, topNameRefs, topSizes, loc);
+  threadsToInpTensor.merge("flatDim", 0, topNameRefs);
+  TransformMapAttr threadsToInpTensorAttr = threadsToInpTensor.get();
+  redInput =
+      rewriter.create<TransformOp>(loc, redInput, threadsToInpTensorAttr);
+
+  threadsToInpTensor =
+      BottomUpTMBuilder::above(threadsToInpTensor, threadsToInpTensorAttr);
+  threadsToInpTensor.pad({"flatDim"},
+                         {0, totalThreads * dataPerThread - topSizeProduct});
+  threadsToInpTensorAttr = threadsToInpTensor.get();
+  redInput =
+      rewriter.create<TransformOp>(loc, redInput, threadsToInpTensorAttr);
+
+  threadsToInpTensor =
+      BottomUpTMBuilder::above(threadsToInpTensor, threadsToInpTensorAttr);
+  threadsToInpTensor.unmerge({"bid", "tid", "iter"}, {0, 1, 2}, "flatDim",
+                             {gridSize, blockSize, dataPerThread});
+  TransformMapAttr tidIterAttr = threadsToInpTensor.get();
+  Value transformed = rewriter.create<TransformOp>(loc, redInput, tidIterAttr);
+
+  return transformed;
+}
+
+static LogicalResult getStoreMethod(ReduceMethod rMethod,
+                                    StoreMethod &stMethod) {
+  if (rMethod == ReduceMethod::Sum) {
+    stMethod = StoreMethod::AtomicAdd;
+    return success();
+  }
+  return failure();
+}
+
+LogicalResult ReduceRewritePattern::matchAndRewrite(
+    ReduceOp op, ReduceOpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  auto loc = op->getLoc();
+  uint64_t redAxis = adaptor.getAxisAttr().getInt();
+  auto gridSize = adaptor.getGridSizeAttr().getInt();
```

Comment:
You can and should just call `op.getGridSize()` - `adaptor` is there specifically for values that might've already been rewritten by other conversion patterns.

Reply:
Thanks for the info. I wasn't getting the reason for the mixed usage of the two.

Reply:
done